### ISSUE TITLE
npm install [package name] removes packages

### ISSUE THREAD

**Issue by zalesky** [npm/npm#17379]
npm: 5.0.3
node: 7.8.0
Steps: 
```
react-native init someProj
cd someProj
npm install react-native-hockeyapp 
```
Actual output:
```
+ react-native-hockeyapp@0.5.1
added 1 package and removed 643 packages in 9.524s
```
Expected result:
npm install should only add this package

**Comment by Salakar** [comment]
Happening for a few of our users who are using v5 also in issue mentioned above.

**Comment by axemclion** [comment]
Looks like this is not just a `react-native` issue. 
`npm init -y && yarn add colors && npm install color-name` actually removes colors and installs color-name.

Used `colors` and `color-name` here in the example since they don't have dependencies. 

As a work around, if you have `yarn` on the machine, React Native installs using yarn. You would need to do `npm install` to generate the lock file, and then install your dependencies.

**Comment by ScottFreeCode** [comment]
Sounds similar to #16853?

**Comment by iarna** [comment]
Closing this as a duplicate of #16853

**Comment by iarna** [comment]
Not really related to 16853. 16853 is about npm pruning known extraneous packages when doing any sort of install.

This issue is about npm, when adding a new dependency, deleting packages that we can't verify as matching your `package.json` but then not replacing them, leaving you with a `node_modules` folder that can only fixed w/ another install. This shows up in particular when folks switch from yarn to npm. (For reference, the npm→yarn direction results in yarn deleting all of the packages installed by npm.)

**Comment by ghost** [comment]
I've had the same issue on Windows 10.

npm - 5.5.1
Node.js - 8.7.0

Console output:
```
C:\path>npm install lodash --save
npm WARN angular-material@1.1.1 requires a peer of angular@>=1.3 <1.6 but none is installed. You must install peer dependencies yourself.
npm WARN angular-material@1.1.1 requires a peer of angular-animate@>=1.3 <1.6 but none is installed. You must install peer dependencies yourself.
npm WARN angular-material@1.1.1 requires a peer of angular-aria@>=1.3 <1.6 but none is installed. You must install peer dependencies yourself.
npm WARN private.package@1.0.0 No repository field.
npm WARN private.package@1.0.0 No license field.

+ lodash@4.17.4
removed 1440 packages and updated 1 package in 9.114s
```

**Comment by egucciar** [comment]
im having this same problem, im playing whack-a-mole, installing one dependency for no reason is causing another to uninstalll. It is happening regardless of id i am passing --save-dev

```
// after running npm install [name]
removed 669 packages and updated 13 packages in 42.024s

// then I ran npm install shortly after:
added 663 packages and updated 12 packages in 140.936s
```

So far i have experienced this in 2 repos. Maybe it has to do with not updating package-lock.json?

Stats:
Win 10
Npm v 5.3.0
Node 6.11.2

**Comment by vith** [comment]
This still happens on `npm 5.6.0` / `node 9.3.0`. It affects packages installed from `git+ssh://` URIs even if they were installed by npm (rather than yarn).

**Comment by RoboPhred** [comment]
Can verify npm 5.6.0.  Using git+http in my case.  This might be a related-but-different issue specifically with failing to verify git packages?

**Comment by narkowicz** [comment]
Also seeing this removing all `git+ssh://` dependencies when installing anything with `5.6.0` in some of my projects.

```
npm install --save fs-extra
+ fs-extra@5.0.0
added 2 packages, removed 363 packages and updated 1 package in 9.783s
```

No trace of `yarn` in these projects, cleaning out `package-lock.json` and `node_modules` doesn't help (beyond effectiveness of initial re-install).

**Comment by eladitzhakian** [comment]
I get this with npm 5.6.0 using `git+ssh` as well

**Comment by GrantMStevens** [comment]
+1

running `npm install` will install packages using `git+ssh` from our package.json just fine, but any subsequent package installs will remove those packages that were installed using `git+ssh`.

**Comment by leoskyrocker** [comment]
+1 what @GrantMStevens says.
```
$ npm --version
5.6.0

$ npm install immutable --save
+ immutable@3.8.2
added 1 package and removed 3 packages in 13.682s

$ npm install
added 3 packages in 6.156s
```

**Comment by clintmod** [comment]
Is this not reproducible? Seems like a pretty serious bug. I lost hours on this.

**Comment by jonathantneal** [comment]
I’m also experiencing this issue, which is not a terribly useful comment, but I’m strategically writing something here to prevent @npm-robot from automatically closing this issue after some arbitrary threshold of inactivity. When I first reported this issue in May of 2017 (#16798), it took 7 days of inactivity.

I wouldn’t want the holidays to be your demise, #17379. 🤗 🎅 

```
npm --version
5.6.0

node --version
v9.3.0
```

**Comment by DaAwesomeP** [comment]
When installing any package, NPM removes my Git packages unless I specifically note to reinstall them in that command.

NodeJS v9.3.0
NPM v5.6.0
Windows 10 Pro x64

**Comment by MatrixSenpai** [comment]
Also seeing this issue. However for me it's outside of a React project, just in a directory with a package created by `npm init`. None of my deps are `dev-dependencies` or saved with `--save-dev`
```
Ubuntu: 17.10
NodeJS: v8.9.1
NPM: v5.6.0

**Comment by vtam311** [comment]
Same issue here as well. 

To reproduce:
-Install a package by adding "package-name": "github:username:package-name" to package.json then npm install.
-Install another package from git, like above.
-Terminal output shows: "added 3 packages and removed 1 package"

Machine:
npm -v: 5.6.0
node -v: 8.9.1

Possibly relevant for reproducing the issue:
The issue hadn't occurred for me since upgrading to the latest react-native version, 0.51.0. 

As a workaround, has anyone tried lowering their npm version on their machine? Not sure if I want to do that considering I have other projects that rely on npm and would prefer consistency across all projects, but am curious if there isn't a solution for this issue soon.

**Comment by saiichihashimoto** [comment]
Same here. Not using any kind of boilerplate or app generator, my github repos get uninstall when I install any new package. I have to rerun `npm install` to bring them back.

**Comment by reblace** [comment]
It happens when installing local files also. Makes it difficult to test packages before you publish them.

e.g.:

```
cd firstProject
npm pack
cd ../secondProject
npm install ../firstProject-1.0.0.tgz
```

**Comment by stouf** [comment]
It happens for me too with dev dependencies installed via github.

**Comment by RoboPhred** [comment]
I believe I am on the trail to the cause of this.  I am not familiar with the workings of npm, but this is what I have discovered so far:

NPM correctly loads the git dependency into its currentTree.  However, cloneCurrentTreeToIdealTree removes it from idealTree because it has neither a package._shasum nor a package._integrity.
It is briefly reintroduced by a package-lock.json, if present.  However, this re-introduced node has an empty array for node.requiredBy, causing it to be removed by pruneIdealTree.  Finally, the dependency is deleted due to its presence in currentTree but not idealTree.

As a bonus: the lack of the git dep in idealTree causes the shrinkwrap logic to not find it as onDisk, resulting in inflatableChild re-fetching the package from the git repo.  I have noticed that all git packages tend to always reinstall regardless if they have changed, and I suspect this is the cause.

**The solution seems to be to have npm write an _integrity or _shasum key to git packages on install**.  I suspect the lack of these might be a bug, but I have not looked into that system yet.


The series of events looks like this:

1. The initial tree is loaded correctly.  My local git dep is loaded and present as a child of the parent project's node by the end of ```normalizeCurrentTree```
2. On loadIdealTree, our **dep is thrown out** by [cloneCurrentTreeToIdealTree](https://github.com/npm/npm/blob/5e426a78ca02d0044f8dd26e0c5f881217081cbd/lib/install.js#L701), because it filters out packages that have neither _integrity or _shasum set.  The git dependency has neither.
3.  npm then applies the shrinkwrap: on ```function inflateShrinkwrap (topPath, tree, swdeps, opts)```, our git dep is not in the idealTree, so it is not loaded into onDisk,
4. Because the dep is not in the onDisk array, inflatableChild has to re-fetch the dep.  This is most likely a **second bug** caused by this bug's interaction with 3, as I have noticed npm will always re-download git repo entries during install ops.
5. Since there is no onDisk entry for our dep, inflatableChild calls fetchChild, which loads the git dep from the dependency.  However, since this was loaded in isolation, the resulting child node **has a ``requiredBy`` of length zero**.
6. loadAllDepsIntoIdealTree calls and completes
7. pruneIdealTree is called.  At the start of the call, this.idealTree has our git dep (as a result of the shrinkwrap load), but it has a userRequired = false and a requiredBy of length zero.
8.  Because the package is not user required nor has any requiredBy entries, isExtranious returns true and pruneIdealTree deletes it from the tree.
9.  The rest of the install logic runs.  The git dependency is eventually uninstalled because it is present in currentTree but removed in idealTree

I tested the above setup with this package.json
```
{
  "name": "test-dir",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@my-pkg-scope/my-pkg": "git+http://my-internal-gogs-repo:3000/my-org/my_repo.git#master",
    "lodash.keys": "^4.2.0",
    "lodash.values": "^4.3.0"
  }
}
```
I first ran "npm i", then tested with the command
"npm i log-level=silly -S lodash.values"

**Comment by RoboPhred** [comment]
I was able to 'fix' the issue, but there are some comments in the code that make me feel like I might be awakening another bug:
```
// no integrity for git deps as integirty hashes are based on the
// tarball and we can't (yet) create consistent tarballs from a stable
// source.
```

My first hint was that the shrinkwrap logic [explicitly ignores git-package integrity](https://github.com/npm/npm/blob/latest/lib/shrinkwrap.js#L123).  However, this is not the whole story, as the integrity key is entirely missing on the package.

After some hunting, I found that this is controlled by pacote, and that pacote also [explicitly disables integrity](https://github.com/zkat/pacote/blob/latest/lib/fetchers/git.js#L115) for git sources.
(Note: It also does this for local sources, which may explain the behavior seen by @reblace in [the above comment](https://github.com/npm/npm/issues/17379#issuecomment-354369401).
(Edit: On closer look, it skips integrity for [local directories](https://github.com/zkat/pacote/blob/latest/lib/fetchers/directory.js#L31), not packed tar.gz packages.  I cannot explain why this would happen for tars.

By removing the git check in shrinkwrap, and the ```_integrity: false``` key in pacote's git logic, I was able to successfully install non-git packages and have the git packages remain untouched.

However, the integrity comment in the shrinkwrap logic still worries me.  I suspect it might be old / out of date, as pacote will happily create a surrogate tar archive and calculate an integrity sum from it.
I was able to force-clear the npm cache and run "npm install" multiple times without it redownloading the git package, so I assume the tar archive is being generated in a stable form and the checksums match up.

Its worth noting that even if npm was not generating stable checksums, I think would still be better off.  Pacote would store the cached checksum, and it would match the _integrity key of the package on disk.  Cache clears or sharing package-lock between devices would still cause a reinstall, but the current behavior of deleting all git packages every install is pretty much the same behavior in the end.
(This is all from about an hour's worth of looking at the code, let me know if I am wrong here).

**Comment by bbottema** [comment]
I'm also having issues with this. Initial `npm i` installs my "git+https" dependency, but subsequent installs removes it again. `npm update` TWICE adds it again.

npm: 5.6.0
node: v6.10.2

**Comment by atuttle** [comment]
Chiming in as another affected user. Same experience as @bbottema. I've started using `npm install thing && npm install` to always add it back. Incredibly annoying.

npm: 5.6.0
node: 8.9.4

**Comment by anirudhmurali** [comment]
Same here, facing the same with npm `5.6.0`!
`added 1 package and removed 1443 packages in 11.634s`

**Comment by bbottema** [comment]
This is rather annoying since some of my dependencies have elaborate post install scripts and fixing my https+git dependencies every time this way is becoming rather tiresome. Already `npm update` is a lot slower than `npm install` and I have to run it twice to get them back.

**Comment by BeckZero** [comment]
same here,  5.6.0

**Comment by danielsogl** [comment]
Im facing the same issue but only with repos added from a git source.

**Comment by gvillenave** [comment]
Same here with 5.6.0, also with packages from git sources

**Comment by RoCat** [comment]
Hello, 
Not sure if it is exactly the same issue but i have similar behavior when I install packages from a git+ssh uri.
example package.json:

```
{
  "name": "whatever",
  "version": "1.30.0",
  "license": "proprietary",
  "dependencies": {
    "async": "2.4.1",
    "private_module": "git+ssh://git@private_repository/private_module.git#master",
  }
}
```

when I "npm i" both modules are installed. then if i do another npm install such as "npm i lodash", the private module (and only this one) is removed.

Hope this could help ;)

**Comment by ghost** [comment]
The same issue with node v6.10.0 and npm 5.6.0
My solution is:
1. Look up which dependencies are removed by "install" or "update" command
2. Install all missing packages via single command

Example:
```
"dependencies": {
    "a": "git:ssh://....",
    "b": "git:ssh://....",
    "c": "git:ssh://....",
   ...
  }
```
Both packages "b" and "c" are missing after I execute "npm install --save a". In this case I execute "npm install --save a b c". As a result, all packages stay in package.json and package-lock.json files

**Comment by leoskyrocker** [comment]
Hi @tasmanianfox 
A simpler solution that works for me is simply run `npm install` followed by `npm install package-name`.
`npm install` seems to work properly.

**Comment by ghost** [comment]
Hi @leoskyrocker ,

I was able to try your solution only once, but looks like it works well, thank you!

**Comment by luig0** [comment]
I am experiencing the same issue -- npm install package removed a previous package that was installed from git
Running npm install immediately after reinstalled the removed packages

NPM: 5.6.0
Node: 8.9.4

**Comment by andreyvolokitin** [comment]
@RoboPhred, any follow up on this?

> I was able to 'fix' the issue

But how to edit npm executable?

**Comment by shufenghua** [comment]
Same issue -- install one git+ssh package will remove other previous installed git+ssh packages.

NPM: 5.6.0
NodeJS: v9.4.0

**Comment by orkisz** [comment]
Same issue for me - installing any package using `npm install <package>` will remove all packages from git.

NPM: 5.6.0
NodeJS: v8.9.4

**Comment by Standard8** [comment]
Hey people, I don't contribute to npm, but I'm watching this issue so I know when it gets fixed.

Adding lots of "Same issue" or "+1" just serves to spam all those watching the issue and waiting. Please don't do it, it doesn't help anyone, I know it can be upsetting that it hasn't been fixed yet, but the thing this really needs is for a contributor or someone to step up and help fix the issue.

**Comment by jonathantneal** [comment]
@Standard8, thanks for watching the thread. Unless you’ve heard otherwise, these _occasional_ comments may help a little. If the developers are away, occupied, or sipping drinks on some island (which is absolutely, totally reasonable and healthy), then @npm-robot might sweep this issue under the proverbial rug before they return. While this issue has received a label, [do note this comment from @npm-robot](https://github.com/npm/npm/issues/16798#issuecomment-309189667).

Again, I’m not certain how @npm-robot works, and the docs don’t mention the robot or the inactivity policy. I’d just feel worse if it were closed, as it did when I reported it once before. 😸

**Comment by grovesNL** [comment]
I can confirm that it happens for me when I have a git dependency as well.

**Comment by Sammiches327** [comment]
I'm working on a Discord Bot and have come across what I think to be along the same lines as everyone else here. 

Steps to reproduce: `npm install discord.js --save` and then follow up with `npm install discord.js-commando --save`. No matter the order, either command deletes the other package if it is installed when running. Perhaps it's a similar directory name bug?

Edit: I am on macOS High Sierra v. 10.13.2 (17C88), running Node v. 9.4.0 and npm v. 5.6.0.

**Comment by unrevised6419** [comment]
Yesterday encountered this issue. Tried to downgrade to all npm versions till 5.0.0, and the issue persists.  
Downgraded to last 4th version, 4.6.1 and is working. 
```
npm -g i npm@4.6.1
```

Tested on:
```
- 2 x Ubuntu 16.04
- macOS High Sierra v10.13.2
```
```
node: 8.9.2
npm: 5.6.0
```

This will give a diff for `lib/install.js` and `lib/install**/*.js` between the issues
```
git diff v4.6.1..v5.0.0 -- 'lib/install/**/*.js' 'lib/install.js' > npm.diff
```

**Comment by unrevised6419** [comment]
I want to ask a question.  
Does people who encounter this does use `create-react-app` ?
It seems this is just a problem of inconsistency between `npm` and `yarn` lock files
For example  `create-react-app` installs depenencies using `yarn` by default, and creates `yarn.lock`
After that you install a dependency using `npm`, and `package-lock.json` get created with just installed extension.

```
# installs all deps using yarn and creates yarn.lock
create-react-app test
cd test
# installs lodash, creates package-lock.json
npm i lodash
npm start // get error
```
`package-lock.json`
```
{
  "name": "test",
  "version": "0.1.0",
  "lockfileVersion": 1,
  "requires": true,
  "dependencies": {
    "lodash": {
      "version": "4.17.4",
      "resolved": "https://registry.npmjs.org/lodash/-/lodash-4.17.4.tgz",
      "integrity": "sha1-eCA6TRwyiuHYbcpkYONptX9AVa4="
    }
  }
}
```
`yarn.lock`
```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


abab@^1.0.3:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/abab/-/abab-1.0.4.tgz#5faad9c2c07f60dd76770f71cf025b62a63cfd4e"
... 6642 rows
```

After creating an app using `create-react-app`,
And if run once `npm install` you're fixed, as `package-lock.json` get generated for all deps.

PS: `create-react-app` in last version added a parameter that allows to install using npm
```
create-react-app app-name --use-npm
```

Please test this.

**Comment by saiichihashimoto** [comment]
@iamandrewluca I do not use `create-react-app` and don't use `yarn`. tbh, not sure what `yarn` is. Still have an issue where when I `npm install <package>`, all my github url packages uninstall themselves.

**Comment by egucciar** [comment]
@iamandrewluca 

it seems the issue for a lot of us has more to do with git/symlink dependencies than it does for discrepancy between yarn and package.json. I for one have not used yarn and experienced this issue.

Also I think that we should probably stop spamming with "me too" type things. A bunch of us will for sure be notified when this issue gets closed, either due to inactivity or otherwise and there will be a flood of us coming in the check what the deal is.

I for one have seen 10 or so comments re-iterating the exact issue we're all having. It's no secret this is a universal issue with this version (v5+) of NPM and particular dependencies.

Anyway, the point being that, it is quite annoying to get constant emails as people come to realize they have the same issue. I'm in agreement that we should try minimizing the "me too" type replies, and rather Thumbs Up existing replies which report the same issue.  

Also it is worth noting this issue infact DOES have a label, and was not updated for over a month in the summer & still is around. :)

**Comment by Butterfly** [comment]
Mid-October 2017, I lost Internet connectivity for 8 days combined. When I FINALLY had consistent access to my web based inbox, there were at least 100 conversation threads caused by the bot auto-closing a *BUNCH* of old, stale issues. That is possibly one aspect that has caused the expressed "nervousness".

Since October, it doesn't happen often, but I *have* seen what appears to be the bot selectively closing issues that have been inactive from between 3 days and maybe 2 weeks old or so. Some of the change there has been because incomplete and non-`npm` issues et al(l) have been being manually closed almost as soon as those are created these days. *yay* :smile:

Happy `npm`'ing! :smile:

**Comment by ljharb** [comment]
@Butterfly the bot will not close labelled issues, i believe.

**Comment by unrevised6419** [comment]
@egucciar I could reproduce this problem only with git without symlinks.  
Don't think this is related to symlinks. `npm link` has his own universe of issues.  
I am not spamming the thread, just want to debug deeper and get into details, and different use cases appear that can help some users, that may have exact use case. Trying to help.

**Comment by egucciar** [comment]
@iamandrewluca  yes i believe you are correct, i got my wires cross as NPM 5+ also introduces the automatic removal of any sym links on every install, similar concept, same version, different issue. Thx.

**Comment by stouf** [comment]
Although the spamming thing definitely seems to be worth being discussed, I think it's a little off topic and a dedicated issue for it would be a better place to discuss it in my opinion 🙂

**Comment by unrevised6419** [comment]
Here git dependencies gets removed from packages because they don't have _integrity or _shasum
https://github.com/npm/npm/blob/0cc9d89ed2d46745f91d746fda9d205fd39d3daa/lib/install.js#L701

**Comment by RoboPhred** [comment]
@iamandrewluca its more complex than that.  Pacote does not generate checksums for git, and npm explicitly ignores package-lock git integrity.  I tracked this down [two weeks ago](https://github.com/npm/npm/issues/17379#issuecomment-354620986), but have yet to hear back from anyone on why this would be the case.  There are comments about being unable to make stable integrity strings for git, but I have not been able to reason this out, and testing seems to indicate it can generate stable checksums just fine.

I really do need more information on how pacote works, or to hand this off to a node dev.  
However, given how long this has been ignored for, I might just make a pull request with the "fix" and leave whatever issues it might cause for the future to deal with.

If you want to play around with a potential fix:
* Change pacote to auto-generate integrity for git deps.  You can do that by [deleting this line](https://github.com/zkat/pacote/blob/latest/lib/fetchers/git.js#L115)
* Stop npm deleting integrity data on loading package-lock.  [Remove this conditional](https://github.com/npm/npm/blob/latest/lib/shrinkwrap.js#L123) (but keep the body)

**Comment by bbottema** [comment]
A package manager that _removes_ packages when _installing_ packages. World upside down, why is this not getting any attention? Issues like this is what gives javascript development a bad rep.

**Comment by Apollon77** [comment]
But this is not only about git. it also breaks in general. I have the "feeling" (and yes only that) that when "npm install new-package" the already installed packages and content of existing package.json is partly ignored and that's why some things from other packages gets removed ... and it is far more then git repos!

**Comment by RoboPhred** [comment]
@Apollon77 The root issue occurs when pacote refuses to return integrity for a package, which occurs on:
* [git](https://github.com/zkat/pacote/blob/latest/lib/fetchers/git.js#L115)
* [git hosts (github and similar)](https://github.com/zkat/pacote/blob/latest/lib/fetchers/hosted.js)
* [local directories](https://github.com/zkat/pacote/blob/latest/lib/fetchers/directory.js#L31)

In theory, it does not affect:
* [tarballs](https://github.com/zkat/pacote/blob/latest/lib/fetchers/file.js#L40)
* [non-npm registries](https://github.com/zkat/pacote/blob/latest/lib/fetchers/remote.js#L23)

Although I have not tested these, and the tarball code is a bit hard to follow.

Have you seen this occur for other types of installs?  Keep in mind that if one of these is the only requester for an npm-registry dependency, that dependency will also be uninstalled.

**Comment by Apollon77** [comment]
I can test again, but most things where I experienced it only "normal" npm registry packages were involved and not git or local stuff. it was more the case that the new installed package was not already in the package.json.
So expectation is that after installing a new package (now with explicit --save) it is added to package.json AND during install also all dependecies from package are still considered. I have more the feeling it's also something with this.

Se also all first comments in this issue where noone wronte about git/local stuff at al

**Comment by ghost** [comment]
@RoboPhred I have exactly this issue. Npm install "another pkg" removes all my github installed pkgs.

**Comment by atuttle** [comment]
Thankfully [it sounds like this issue is getting some attention](https://github.com/zkat/pacote/issues/133#issuecomment-358932501).

**Comment by robotlolita** [comment]
So, from Kat's comment that @atuttle linked, you should be able to workaround this issue for now by putting the resolved Git URL (which you'll find on `package-lock.json`) in your `package.json`.

**Comment by andreyvolokitin** [comment]
Nope, it can not parse ```package.json``` now: 
```bash
npm ERR! code EJSONPARSE
npm ERR! Failed to parse json
npm ERR! Unexpected token / in JSON at position 1185 while parsing near '...2d0bfed7ce3935689",
```

**Comment by andreyvolokitin** [comment]
No, my bad, it actually works. Thanks @robotlolita!

**Comment by donhamiltoniii** [comment]
Also getting this issue. Temporarily solved by downgrading to `npm 4.6.1`. Obviously not a sustainable fix. Hopefully, we get a fix on this soon.

**Comment by dontsave** [comment]
Happening for me also in 5.6. All git+ssh dependencies are removed every single time npm i is called

**Comment by znorman-harris** [comment]
Having the same problem with 5.6 uninstalling dependencies from GitHub when adding a new package.

**Comment by atuttle** [comment]
# Work-around:

The problem is that you've added a git dep, possibly using a branch identifier like: `git+https://..../you/your_project.git#your_branch` but after resolution it's saved in package-lock.json not as `#your_branch` but as `#sha_for_latest_commit_to_your_branch`. When npm tries to resolve this difference it gets confused and removes what you've got currently.

You can get around the npm bug, while we wait for a fix to land, by copying that sha from package-lock.json into your package.json. You'll need to change the sha in package.json any time the dep gets more commit(s) pushed that you want in your project.

This would be annoying if you wanted it to automatically pick up changes to a frequently changing git dependency, but at least it would stop the uninstall behavior.

**Comment by RoboPhred** [comment]
The issue is unrelated to how package-lock saves.  Adding the shasum to the package.json fixes it because node will no longer try to purge it for being an "old install" (pre-5.x before shasums were introduced).

the package.json is not getting the shasum in the first place because pacote declines to generate an integrity string for any git or local directory dependency.

**Comment by ShadSterling** [comment]
In my case, it looks like `cordova prepare` was uninstalling certain packages.  Inserting another `npm i` between prepare and build made it work.  (But hasn't solve all my problems, because other cordova commands are still unreliable.)

**Comment by scottieslg** [comment]
I'm seeing this problem as well.  I run npm update, and suddenly my angular app breaks because of missing packages.  I have to run npm install after the update to get them back.

**Comment by tlaak** [comment]
I'm also seeing this problem in an Angular 5 project. When I install a new dependency `npm i [ModuleA] -S`, another unrelated ModuleB gets removed. The ModuleB stays in package.json, but it is gone from node_modules and removed from package-lock.json with all its dependencies.

In my case ModuleB points directly to a GitHub repository. I haven't seen this happening with any other packages in my project.

After adding the commit hash, as mentioned in @atuttle's workaround, the problem is gone.

**Comment by gaearon** [comment]
>Does people who encounter this does use create-react-app ?

Yes, one of the possible ways the issue occurs is when you installed the previous deps with Yarn (which CRA does for you if you have it on your system).

What happens is that Yarn doesn’t include integrity metadata into downloaded package.json. (It wasn’t necessary until npm 5, and is not documented either.) However, npm just blows away packages without metadata instead of trying to reinstall them by itself. This is why **running npm install after yarn will always produce this issue**.

It could be fixed on yarn side by writing that metadata: https://github.com/yarnpkg/yarn/issues/5240. Or I suppose it could be fixed on the npm side by reinstalling packages it “doesn’t trust” instead of just deleting them. Hopefully this gets resolved one way or the other!

**Comment by babeal** [comment]
I am seeing something like this with npm update.  If I run npm install and then npm update on 5 of our packages the package.json and the package-lock.json get updated but one of the packages is removed from the node_modules folder.  I don't use yarn and this is on both a build server starting from a clean folder and on all our dev machines.  Something isn't right.

**Comment by knight9999** [comment]
Thanks  @RoboPhred  I found that by using NPM 5.3.0, _integrity is added in package.json for (git) url installed package.  But by using NPM 5.6.0, _integrity is not added for (git) url installed package. Then the latter removes (git) url installed pacakges when installing other package. I wonder why this is developed.
Remark after changing npm version, we must clear npm cache.

**Comment by ctsstc** [comment]
Thought this was an issue with private git repos, but seems to be all git repos.
When adding a new package it removes all my packages that were added via git. Packages are only removed in the `node_modules` directory, the `package.json` still has the package in it.

I then have to run `npm install` to get the removed packages from the node_modules folder back... : \

I'm running npm version: `5.6.0`

**Comment by glutaminefree** [comment]
@ctsstc Same here. Version of **npm** is 5.6.0 too.

**Comment by willbattel** [comment]
What is most strange to me is the inconsistency I'm seeing. Sometimes it works and sometimes it doesn't- and I can't find any pattern between them.

**Comment by donhamiltoniii** [comment]
I'm curious how many converts to yarn this is causing.

**Comment by znorman-harris** [comment]
@donhamiltoniii that is how I "solved" this problem

**Comment by FireController1847** [comment]
Still an issue and a pretty big one at that using v5.6.0

**Comment by stuckless** [comment]
I'll add on to this...

my `package.json`
```
{
  "name": "sagetv-ui",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "bootstrap": "^4.0.0",
    "sagetv-js-api": "git+https://github.com/stuckless/sagetv-js-api.git"
  },
  "scripts": {}
}
```
Run npm install to bring in the packages...
```
$ npm install

npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN bootstrap@4.0.0 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.0.0 requires a peer of popper.js@^1.12.9 but none is installed. You must install peer dependencies yourself.

added 2 packages in 1.227s
```
Check node_modules... looks good, `sagetv-js-api` is there
```
$ npm install jsonpath --save

> jsonpath@1.0.0 postinstall /home/seans/tmp/npmtest/node_modules/jsonpath
> node lib/aesprim.js > generated/aesprim-browser.js

npm WARN bootstrap@4.0.0 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.0.0 requires a peer of popper.js@^1.12.9 but none is installed. You must install peer dependencies yourself.

+ jsonpath@1.0.0
added 30 packages and removed 1 package in 2.047s
```
You can see it removed 1 package.  Checking node_modules shows that `sagetv-js-api` was removed.

```
$ npm update

npm WARN bootstrap@4.0.0 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.0.0 requires a peer of popper.js@^1.12.9 but none is installed. You must install peer dependencies yourself.

+ sagetv-js-api@9.0.0
added 1 package in 1.016s
```
You can see `sagetv-js-api` is brought back.

Kind of annoying to have to keep running update to recover the removed packages after installing a new package :(

This is npm version 5.6.0 on linux.

**Comment by miazbikowski** [comment]
Hi! I'm trying to learn react/JS and this issue's describing the problem I'm hitting. I'm creating my react app then trying to install another library, and it fucks everything up by deleting a ton of modules.

Step 1 I did:
```
npm install -g create-react-app
create-react-app my-app
```

Step 2:
```
npm install recharts
```

which does:
```
+ recharts@1.0.0-beta.10
added 19 packages, removed 1207 packages and updated 24 packages in 11.554s
```

Any guidance for this noob to prevent stuff from getting wiped?

**Comment by stouf** [comment]
@TempestMia The packages that get removed are referenced in your `package.json` using the address of their git repository I imagine? If so, what seems to be the most popular workaround so far is to use the commit hash referencing the version you wish to install, as described [there](https://github.com/npm/npm/issues/17379#issuecomment-359832471) 😉

**Comment by nbolender** [comment]
@TempestMia You could also switch to yarn, which is compatible with npm and doesn’t have this stupid bug.

**Comment by miazbikowski** [comment]
Thanks @stouf - but I hadn't even created a repo for my project yet, I don't seem to have a `package-lock.json`.

 And thanks @nbolender - yours got a lot of thumbs down, are people not fans of yarn?

**Comment by stouf** [comment]
> but I hadn't even created a repo for my project yet, I don't seem to have a `package-lock.json`.

Right. I just tried and even if you use the commit hash workaround, the issue will still occur if you don't have a `package-lock.json` file apparently (I imagine it will be the same with `npm-shrinkwrap.json`, I haven't tried though).
I would say that your best option at the moment is to use both the `package-lock.json` and the commit hash of the packages you install through the address of their git repository?

> And thanks @nbolender - yours got a lot of thumbs down, are people not fans of yarn?

I'm quite sure it's more about the comment itself not helping the issue than about yarn. I haven't tried Yarn myself but I heard a lot of positive feedback about it :)

**Comment by miazbikowski** [comment]
@stouf @nbolender I ended up using yarn and it worked :)  thank you both!

**Comment by justin-m-lacy** [comment]
@nbolender Will yarn be able to find npm packages or do they manage completely separate repositories?

**Comment by stouf** [comment]
@lerpinglemur Yarn has been designed to be compatible with NPM so there should theoretically be no problems for finding your favorite packages ;-)

Let's close the topic about Yarn or other alternatives if you people don't mind, just for the sake of limiting the content of this issue to its actual topic :)

**Comment by BerndWessels** [comment]
I am having this issue too on `npm @ 5.6.0` on Windows 10.

Especially with dependencies coming directly from Git repos like

```
  "devDependencies": {
    "node-sass": "^4.7.2",
    "react-app-rewire-css-modules": "github:berndwessels/react-app-rewire-css-modules",
    "react-app-rewire-hot-loader": "^1.0.0",
    "react-app-rewired": "^1.4.0",
    "react-hot-loader": "^3.1.3",
    "react-scripts": "1.1.0",
    "sass-loader": "^6.0.6"
  },
```

everytime I add a new dependency with `npm install --save something` the `react-app-rewire-css-modules` disappears.

I consider that a major bug in npm. Or at least something very very annoying that will cause many wasted developer hours and a loss of confidence in npm.

**Comment by josnidhin** [comment]
I am experiencing this issue when it comes to dependencies installed using `git+ssh://` url. 
I have been following this issue here for sometime now. Is there an official timeline regarding a fix for this issue. 

This is a serious problem especially for NodeJS beginners. I have personally seen beginners banging their heads because of this bug.

**Comment by d3x0r** [comment]
I had work in progress in the package it deleted when I installed another unrelated package.
Lost hours of work.......I had commited most of it; but it wasn't pushed (apparently). Edit : friggin dayz OMG.  I had installed the package with --save in the first place so it should have been tracked anyway... but it wasn't building directly.  So I had --force, and was working on it...

**Comment by jbrodriguez** [comment]
I'll say just one thing ... this is a **painful** error.

**Comment by joshdick** [comment]
npm 5.7 had [several pruning bugfixes](http://blog.npmjs.org/post/171139955345/v570), maybe it fixes this issue?

**Comment by zkat** [comment]
This is fixed as of `npm@5.7.1`
